src/main/java/com/google/devtools/build/lib/skyframe/RepositoryMappingFunction.java

```diff
@@ -135,6 +135,10 @@ public SkyValue compute(SkyKey skyKey, Environment env)
         return computeForModuleExtensionRepo(
             repositoryName, moduleExtensionId.get(), extensionEvalValue, bazelDepGraphValue);
       }
+
+      if (RepositoryName.BAZEL_TOOLS.equals(repositoryName)) {
```

Comment:
What is the justification for this check here?

The complexity of the code here is going slightly past my review capabilities and my first thought is this is just making the tests to pass and the code even more complex.

Similar complexity to this one is happening in BzlloadFunction.getRepositoryMapping, which I imagine is calling the location here.

Is there a way to simplify it?

Reply:
I also don't understand this check. `@bazel_tools` should have its repo mapping calculated just like any other Bazel module, not the `VALUE_FOR_ROOT_MODULE_WITHOUT_REPOS` sentinel value.

> Similar complexity to this one is happening in BzlloadFunction.getRepositoryMapping, which I imagine is calling the location here.

Indeed, `BzlLoadFunction#getRepositoryMapping` calls to `RepositoryMappingFunction`, but the former is more complex because repo mapping computation for .bzl files has extra considerations due to some "partial" loading. `RepositoryMappingFunction` deals with the general case of "all WORKSPACE/Bzlmod/injection has finished and we're just loading a normal BUILD file (and similar environments) now".

Reply:
This check avoids a fundamental cycle when bzlmod is enabled: workspace loading -> `@_builtins` loading -> `@_builtins` repository mapping -> `@bazel_tools` repository mapping -> `//external` package -> workspace loading.

The effect of this check, I believe, is you can't declare a repo_mapping for `@bazel_tools` in WORKSPACE. That's fine, I think? I'm certainly open to any other ideas on how to resolve this cycle.

Reply:
The conditions that lead to this line are (I took the liberty to enumerate them using and IDE, because I find it hard to follow {,} on github):
- enableBzlMod 
- not @_builtins
- not (main repo && rootModuleShouldSeeWorkspaceRepos)
- bazelDepGraphValue.getCanonicalRepoNameLookup().get(repositoryName) == null
- maybeGetModuleExtensionForRepo(repositoryName) == null

The conditions seem mostly fine, providing we can't use `bazel_tools` for either a module name or module extension within bzlmod. Is this true?
In case it can be used, you're back forming a cycle as soon as the new `bazel_tools` uses `builtins`. Or even worse, you don't allow this new repo to be used from anywhere (even when it should be locally available).

What I'm also concerned is, this doesn't seem to handle the same cycle when `not enableBzlMod`.

Also, because the cycle is started from @_builtins repository mapping request. I think this would be better put into `BzlloadFunction.getRepositoryMapping`, so that it covers only @_builtins -> @bazel_tools edge.

In that function it might even be clearer which cases need cycle-breaking.

(I'm not an expert in SkyFrame, I've written what I think and not everything might be completely accurate)

Reply:
I'd also suggest adding a test for the minimal reproduction in the bug, having bzlmod enabled and disabled.

Reply:
Hang on, I still don't understand. the branch on line 119 should include `@bazel_tools`, as `bazel_tools` is a builtin module ("builtin" as in "shipped with Bazel", not as in `@_builtins`). So this check on line 139 should never occur.

(In other words, from what Ivo wrote, the fourth bullet point should not hold.)

Alternatively, the question is -- how come this cycle wasn't a problem before? How does injection builtins for WORKSPACE-loaded .bzl files affect the repo mapping calculation of `@bazel_tools` when Bzlmod is enabled?

Reply:
This looks good to me now. @Wyverald any concerns?